File: backend/routes/problem.js
const Router = require("koa-router");
const ProblemModel = require("../databases/models/problem");
const router = new Router();
const { resolveAuthorizationHeader } = require("../utils");
const jsonwebtoken = require("jsonwebtoken");

router.get("/", async (ctx) => {
  try {
    const res = await ProblemModel.find();

    ctx.body = {
      data: res,
    };
  } catch (error) {
    ctx.body = {
      message: error,
    };
  }
});

router.get("/:id", async (ctx) => {
  const params = ctx.params;
  const { id } = params;
  try {
    const res = await ProblemModel.findOne({ _id: id });

    ctx.body = res;
  } catch (error) {
    ctx.body = {
      message: error,
    };
  }
});

router.post("/", async (ctx) => {
  const body = ctx.request.body;
  const { title, description, answer, difficulty, status } = body;
  const token = resolveAuthorizationHeader(ctx);
  const user = jsonwebtoken.decode(token);
  const problem = new ProblemModel({
    title,
    description,
    answer,
    difficulty,
    status,
    createUserId: user.id,
  });

  try {
    const res = await problem.save();

    if (res) {
      ctx.body = { message: "问题创建成功" };
    } else {
      ctx.body = { message: "问题创建失败" };
    }
  } catch (error) {
    ctx.body = {
      message: error,
    };
  }
});

router.put("/:id", async (ctx) => {
  const params = ctx.params;
  const body = ctx.request.body;
  const { id } = params;
  const { title, description, answer, difficulty, status } = body;

  try {
    const problem = await ProblemModel.findById(id, {
      title: 1,
      description: 1,
      answer: 1,
      difficulty: 1,
      status: 1,
    });

    problem.title = title;
    problem.description = description;
    problem.answer = answer;
    problem.difficulty = difficulty;
    problem.status = status;

    const res = await problem.save();

    if (res) {
      ctx.body = { message: "问题更新成功" };
    } else {
      ctx.body = { message: "问题更新失败" };
    }
  } catch (error) {
    ctx.body = {
      message: error,
    };
  }
});

module.exports = router;

File: backend/routes/problem.go
package routes

import (
	"backend/models"
	"strconv"

	"github.com/kataras/iris/v12"
	"github.com/kataras/iris/v12/middleware/jwt"
)

func GetProblems(ctx iris.Context) {
	status := ctx.URLParam("status")

	problems, err := models.GetProblems(models.StatusType(status))

	if err != nil {
		ctx.StopWithProblem(iris.StatusInternalServerError, iris.NewProblem().Title("查询问题失败").Detail(err.Error()).Type("Select Problem"))
		return
	}

	ctx.JSON(Success{
		Code: 0,
		Msg:  "success",
		Data: problems,
	})
}

func GetProblem(ctx iris.Context) {
	// id, err := strconv.Atoi(ctx.Params().Get("id"))

	// ctx.JSON(Success{
	// 	Code: 0,
	// 	Msg:  "success",
	// 	Data: ,
	// })
}

func AddProblem(ctx iris.Context) {
	claims := jwt.Get(ctx).(*UserClaims)
	userId := claims.UserID

	var problem models.CreateProblemItem

	err := ctx.ReadJSON(&problem)

	if err != nil {
		ctx.StopWithProblem(iris.StatusBadRequest, iris.NewProblem().Title("问题参数错误").Detail(err.Error()).Type("Param Problem"))
		return
	}

	err = models.AddProblem(problem.Name, problem.Description, problem.Answer, problem.Difficulty, problem.Types, userId)

	if err != nil {
		ctx.StopWithProblem(iris.StatusInternalServerError, iris.NewProblem().Title("创建问题失败").Detail(err.Error()).Type("Insert Problem"))
		return
	}

	ctx.JSON(Success{
		Code: 0,
		Msg:  "success",
	})
}

func GetProblemsByUserId(ctx iris.Context) {
	claims := jwt.Get(ctx).(*UserClaims)
	userId := claims.UserID

	problems, err := models.GetProblemsByUserId(userId)

	if err != nil {
		ctx.StopWithProblem(iris.StatusInternalServerError, iris.NewProblem().Title("查询问题失败").Detail(err.Error()).Type("Select Problem"))
		return
	}

	ctx.JSON(Success{
		Code: 0,
		Msg:  "success",
		Data: problems,
	})
}

func GetProblemTypes(ctx iris.Context) {
	status := ctx.URLParam("status")

	problemTypes, err := models.GetProblemTypes(models.StatusType(status))

	if err != nil {
		ctx.StopWithProblem(iris.StatusInternalServerError, iris.NewProblem().Title("查询问题失败").Detail(err.Error()).Type("Select Problem"))
		return
	}

	ctx.JSON(Success{
		Code: 0,
		Msg:  "success",
		Data: problemTypes,
	})

}

func UpdateProblem(ctx iris.Context) {
	claims := jwt.Get(ctx).(*UserClaims)
	userId := claims.UserID
	id, err := strconv.ParseInt(ctx.Params().Get("id"), 10, 32)

	if err != nil {
		ctx.StopWithProblem(iris.StatusBadRequest, iris.NewProblem().Title("问题 ID 错误").Detail(err.Error()).Type("Param Problem"))
		return
	}

	var problem models.UpdateProblemItem

	err = ctx.ReadJSON(&problem)

	if err != nil {
		ctx.StopWithProblem(iris.StatusBadRequest, iris.NewProblem().Title("问题参数错误").Detail(err.Error()).Type("Param Problem"))
		return
	}

	err = models.UpdateProblem(int32(id), problem, userId)

	if err != nil {
		ctx.StopWithProblem(iris.StatusInternalServerError, iris.NewProblem().Title("更新问题失败").Detail(err.Error()).Type("Update Problem"))
		return
	}

	ctx.JSON(Success{
		Code: 0,
		Msg:  "success",
	})
}
